text.go

```diff
@@ -13,10 +15,160 @@ const (
 	TextAlignTrailing
 )
 
+// TextWrap represents how text longer than the widget's width will be wrapped.
+type TextWrap int
+
+const (
+	// TextWrapOff extends the widget's width to fit the text, no wrapping is applied.
+	TextWrapOff TextWrap = iota
+	// TextTruncate trims the text to the widget's width, no wrapping is applied.
+	TextTruncate
+	// TextWrapBreak trims the line of characters to the widget's width adding the excess as new line.
+	TextWrapBreak
+	// TextWrapWord trims the line of words to the widget's width adding the excess as new line.
+	TextWrapWord
+)
+
 // TextStyle represents the styles that can be applied to a text canvas object
 // or text based widget.
 type TextStyle struct {
 	Bold      bool // Should text be bold
 	Italic    bool // Should text be italic
 	Monospace bool // Use the system monospace font instead of regular
 }
+
+// MeasureText uses the current driver to calculate the size of text when rendered.
+func MeasureText(text string, size int, style TextStyle) Size {
+	return CurrentApp().Driver().RenderedTextSize(text, size, style)
+}
+
+// SplitLines accepts a slice of runes and returns a slice containing the
+// start and end indicies of each line delimited by the newline character.
+func SplitLines(text []rune) [][2]int {
+	var low, high int
+	var lines [][2]int
+	length := len(text)
+	for i := 0; i < length; i++ {
+		c := text[i]
+		if c == '\r' {
+			// Ignore Carriage Return unless followed by New Line
+			i++
+			if i < length {
+				c = text[i]
+			} else {
+				break
+			}
+		}
+		if c == '\n' {
+			high = i
+			lines = append(lines, [2]int{low, high})
+			low = i + 1
+		}
+	}
+	return append(lines, [2]int{low, length})
+}
+
+// LastWordBoundary accepts a slice of runes and returns the last index of a word delimiter.
+func LastWordBoundary(text []rune) int {
+	i := len(text) - 1
+	for ; i >= 0; i-- {
+		if unicode.IsSpace(text[i]) {
+			break
+		}
+	}
+	return i
+}
+
+// LineBounds accepts a slice of runes, a wrapping mode, a maximum line width and a function to measure line width.
+// LineBounds returns a slice containing the start and end indicies of each line with the given wrapping applied.
+func LineBounds(text []rune, wrap TextWrap, maxWidth int, measurer func([]rune) int) [][2]int {
+	var bounds [][2]int
+	lines := SplitLines(text)
+	switch wrap {
+	case TextTruncate:
+		if maxWidth > 0 {
```

Comment:
we prefer to exit early on empty cases so that the "main business" of a function has as little nesting as possible.
I think `if maxWidth == 0 { return lines }` helps to reduce complexity here.

Reply:
Ack